### PR TITLE
perf: use subarray instead of slice in dispatch minimal

### PR DESCRIPTION
**Pull request by ry**
attempt 2 #4173

### CODE REVIEW

**Comment by bartlomieju** [comment]
`f560617` updates shared queue to transparently align all records to 4 byte words. Problem described in #4173 is now gone.

File: core/shared_queue.js [hunk 1]
```diff
@@ -115,7 +115,13 @@ SharedQueue Binary Layout
...
         return HEAD_INIT;
       } else {
-        return shared32[INDEX_OFFSETS + 2 * (index - 1)];
+        const prevEnd = shared32[INDEX_OFFSETS + 2 * (index - 1)];
+        const prevModulo = prevEnd % 4;
+        if (prevModulo === 0) {
+          return prevEnd;
+        } else {
+          return prevEnd + (4 - prevModulo);
+        }
```
Comment (ry) [inline core/shared_queue.js]: return (prevEnd + 3) & ~3;

File: core/shared_queue.rs [hunk 2]
```diff
@@ -134,7 +134,13 @@ impl SharedQueue {
...
       } else {
         let s = self.as_u32_slice();
-        s[INDEX_OFFSETS + 2 * (index - 1)] as usize
+        let prev_end = s[INDEX_OFFSETS + 2 * (index - 1)] as usize;
+        let prev_modulo = prev_end % 4;
+        if prev_modulo == 0 {
+          prev_end
+        } else {
+          prev_end + (4 - prev_modulo)
+        }
```
Comment (ry) [inline core/shared_queue.rs]: Here too

File: core/shared_queue.rs [hunk 3]
```diff
@@ -169,29 +175,38 @@ impl SharedQueue {
...
   pub fn push(&mut self, op_id: OpId, record: &[u8]) -> bool {
     let off = self.head();
+    assert_eq!(off % 4, 0);
     let end = off + record.len();
+    let len_modulo = record.len() % 4;
+    let aligned_end = if len_modulo == 0 {
+      end
+    } else {
+      end + (4 - len_modulo)
+    };
```
Comment (ry) [inline core/shared_queue.rs]: And here

File: core/shared_queue.js [hunk 4]
```diff
@@ -125,16 +126,19 @@ SharedQueue Binary Layout
   function push(opId, buf) {
     const off = head();
     const end = off + buf.byteLength;
+    const lenModulo = buf.byteLength % 4;
+    const alignedEnd = lenModulo === 0 ? end : end + (4 - lenModulo);
```
Comment (ry) [inline core/shared_queue.js]: const alignedEnd = (end + 3) & ~3

File: core/shared_queue.rs [hunk 5]
```diff
@@ -169,29 +170,34 @@ impl SharedQueue {
...
 
-  /// Because JS-side may cast `record` to Int32Array it is required
-  /// that `record`'s length is divisible by 4.
+  /// Because JS-side may cast popped message to Int32Array it is required
+  /// that every message is aligned to 4-bytes.
   pub fn push(&mut self, op_id: OpId, record: &[u8]) -> bool {
     let off = self.head();
+    assert_eq!(off % 4, 0);
     let end = off + record.len();
+    let len_modulo = record.len() % 4;
```
Comment (ry) [inline core/shared_queue.rs]: unused?

File: core/shared_queue.rs [hunk 6]
```diff
@@ -153,7 +154,7 @@ impl SharedQueue {
 
     let off = self.get_offset(i).unwrap();
     let (op_id, end) = self.get_meta(i).unwrap();
-
+    println!("off {} end {}", off, end);
```
Comment (piscisaureus) [inline core/shared_queue.rs]: Remove?

**Pull request merged**